### PR TITLE
CUSTCOM-12 Fixed node configuration in admin UI

### PR DESCRIPTION
**Pull request by dmatej**
# Description
This is a bug fix for incorrect GUI behavior when configuring Nodes, especially SSH.

* fixed logging of errors (did not log anything in some types of failures)
* fixed distribution of error messages to the user (I still don't like it's formatting, but that's not a thing of this issue)
* fixed updating SSH Node connector when moving from password to keyfile and vice-versa

* validation before the update ends with verification that the keyfile exists. It does not test it's content. Originally it merged defaults, existing configuration and new configuration, so finally it ended with several kinds of exceptions (some of them were swallowed, so GUI showed only "See server.log", but there was nothing).

# Important Info

# Testing

### New tests

* no new automatic tests

#### Manual

* manual testing GUI, looking on changes in server.log and domain.xml
* create SSH Node, update SSH configuration, ping
* create DCOM node, update it's configuration, ping

```
asadmin create-node-dcom --force=true --nodehost=nodehost123 --passwordfile=./passwordfile-windows.txt  cmd-dcom001
Warning: some parameters appear to be invalid.
Unknown host nodehost123
Continuing with node creation due to use of --force.
Command create-node-dcom executed successfully.

asadmin create-node-ssh --force=true --nodehost=localhost1 cmd-ssh001
Warning: some parameters appear to be invalid.
Unknown host localhost1
Continuing with node creation due to use of --force.
Command create-node-ssh executed successfully.

asadmin create-node-ssh --force=true --nodehost=localhost --sshkeyfile=./sss cmd-ssh002
Warning: some parameters appear to be invalid.
Key file path ./sss must be an absolute path.
Continuing with node creation due to use of --force.
Command create-node-ssh executed successfully.

asadmin create-node-ssh --force=true --nodehost=localhost --sshkeyfile=/app/appservers/passwords/payara-test-id_rsa cmd-ssh002
remote failure: Configuration not added. A Node instance with a "cmd-ssh002" name already exists in the configuration
Command create-node-ssh failed.

asadmin create-node-ssh --force=true --nodehost=localhost --sshkeyfile=/app/appservers/passwords/payara-test-id_rsa cmd-ssh003
Command create-node-ssh executed successfully.

asadmin update-node-ssh --force=true --nodehost=localhost --sshport=32000 cmd-ssh003
Command update-node-ssh executed successfully.

asadmin update-node-ssh --force=true --nodehost=localhost --sshport=22 --sshkeyfile=/app/appservers/passwords/payara-test-id_rsad cmd-ssh003
Warning: some parameters appear to be invalid.
Key file /app/appservers/passwords/payara-test-id_rsad not found. The key file path must be reachable by the DAS.
Continuing with node update due to use of --force.
Command update-node-ssh executed successfully.

asadmin update-node-ssh --force=true --nodehost=localhost --sshport=22 --sshkeyfile=/app/appservers/passwords/payara-test-id_rsa cmd-ssh003
Command update-node-ssh executed successfully.

sadmin update-node-ssh --force=true --nodehost=localhost --sshport=22 --sshkeyfile=/app/appservers/passwords/payara-test-id_rsa --sshuser=payara-test cmd-ssh003
Command update-node-ssh executed successfully.
```

### Test suites executed

- Quicklook - NO
- Payara Samples - **OK**
- Java EE7 Samples - NO
- Java EE8 Samples - NO
- Payara Private Tests - NO
- Payara Microprofile TCKs Runner - NO
- Jakarta TCKs - NO
- Mojarra - NO
- Cargo Tracker - NO

### CODE REVIEW

**Comment by dmatej** [comment]
Jenkins test please

**Comment by dmatej** [comment]
Blah, I can only confirm it ... seems I lost something in cherrypicks, because I am sure it worked in original branch :(

**Comment by dmatej** [comment]
I deleted my last comment, because when I was testing now, I forgot to press Save button :D
On the seccond attempt, both in this and second branch where these commits are, it works. 
Can you record a video to see what you did? Or try it again, at least?

**Comment by dmatej** [comment]
But I was able to break it again with know exception with unknown cause ... I think it should be in separate issue and at this moment I don't understand what exactly causes it. (unresolved log message is targeted under CUSTCOM-55).
It broke when I changed user from payara-test to dmatej and keyfile from  to ${user.home}/.ssh/id_rsa . Restart helped.
```
[#|2020-03-18T13:49:34.717+0100|WARNING|Payara 5.202|javax.enterprise.system.core|_ThreadID=185;_ThreadName=admin-thread-pool::admin-listener(2);_TimeMillis=1584535774717;_LevelValue=900;|
  failed.to.update.node ssh003
org.jvnet.hk2.config.TransactionFailure: This service has been unbound: SystemDescriptor(
        implementation=org.glassfish.config.support.GlassFishConfigBean
        name=com.sun.enterprise.config.serverbeans.SshAuth
        contracts={com.sun.enterprise.config.serverbeans.SshAuth,org.jvnet.hk2.config.ConfigBean,org.jvnet.hk2.config.ConfigBeanProxy}
        scope=javax.inject.Singleton
        qualifiers={javax.inject.Named}
        descriptorType=CLASS
        descriptorVisibility=NORMAL
        metadata=
        rank=0
        loader=org.jvnet.hk2.config.ConfigModel$SafeHk2Loader@2d658712
        proxiable=null
        proxyForSameScope=null
        analysisName=null
        id=1700
        locatorId=0
        identityHashCode=1554410178
        reified=true)
        at org.jvnet.hk2.config.WriteableView.commit(WriteableView.java:408)
        at org.jvnet.hk2.config.Transaction.commit(Transaction.java:114)
        at org.jvnet.hk2.config.ConfigSupport._apply(ConfigSupport.java:182)
        at org.jvnet.hk2.config.ConfigSupport.apply(ConfigSupport.java:139)
        at org.jvnet.hk2.config.ConfigSupport.apply(ConfigSupport.java:118)
        at com.sun.enterprise.v3.admin.cluster.UpdateNodeCommand.updateNodeElement(UpdateNodeCommand.java:222)
```

**Comment by dmatej** [comment]
I cannot reproduce it any more after the rebase to current master

**Comment by dmatej** [comment]
Jenkins test please

**Comment by dmatej** [comment]
Jenkins test please

**Comment by dmatej** [comment]
Aha, that's because you did not provide the key file. The problem is that empty value cannot be replaced by default, because it would go against asadmin command which does not change what is not set (reason why it originally did not work at all).
Maybe if it would be resolved in UI layer somehow ...

**Comment by dmatej** [comment]
Jenkins test please

File: nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeRemoteCommand.java [hunk 1]
```diff
@@ -203,21 +203,28 @@ private ParameterMap createValidationParameters(final Node node) {
...
+                    sshkeypassphrase, getSupplier(ssha, ssha::getKeyPassphrase));
+            }
         } else {
-            // if keyfile is set, don't configure password.
-            parameters.insert(NodeUtils.PARAM_SSHKEYFILE, sshkeyfile, getSupplier(ssha, ssha::getKeyfile));
-            parameters.insert(NodeUtils.PARAM_SSHKEYPASSPHRASE, //
-                sshkeypassphrase, getSupplier(ssha, ssha::getKeyPassphrase));
+            if (SshAuthType.KEY.name().equals(sshAuthType)) {
+                parameters.insert(NodeUtils.PARAM_SSHKEYFILE, sshkeyfile, getSupplier(ssha, ssha::getKeyfile));
+                parameters.insert(NodeUtils.PARAM_SSHKEYPASSPHRASE, //
```
Comment (Pandrex247) [inline nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeRemoteCommand.java]:
```suggestion
                parameters.insert(NodeUtils.PARAM_SSHKEYPASSPHRASE, 
```

**Comment by Pandrex247** [comment]
Admin GUI seems to work nicely now.

It might be worth hiding the new `--sshauthtype` parameter for the asadmin commands, since it's not really usable from the command line - you aren't allowed to specify a password or alias from the command line so essentially all the parameter does is wipe your config 😂

Speaking of which, I see that it clears the key and password config when you switch between them.
Just double-checking - this isn't going to trip anyone up is it? To me it makes sense but I don't know if there's any sort of "fallback" behaviour typically expected from SSH.

File: nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeSshCommand.java [hunk 2]
```diff
@@ -72,6 +73,9 @@
     private String sshportInSubClass;
     @Param(name = "sshuser", optional = true)
     private String sshuserInSubClass;
+    /** {@link SshAuthType} name */
+    @Param(name = "sshauthtype", optional=true)
```
[the remark this answers is not in the document]
Comment (dmatej) [reply nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeSshCommand.java]: Copy paste effect from another class :dancer: 
I agree :-)

File: nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeRemoteCommand.java [hunk 3]
```diff
@@ -203,21 +203,28 @@ private ParameterMap createValidationParameters(final Node node) {
...
+        if (sshAuthType == null) {
+            if (sshkeyfile == null && remotepassword == null) {
+                parameters.insert(NodeUtils.PARAM_SSHPASSWORD, null, getSupplier(ssha, ssha::getPassword));
+                parameters.insert(NodeUtils.PARAM_SSHKEYFILE, null, getSupplier(ssha, ssha::getKeyfile));
+                parameters.insert(NodeUtils.PARAM_SSHKEYPASSPHRASE, null, getSupplier(ssha, ssha::getKeyPassphrase));
+            } else if (remotepassword != null) {
+                parameters.insert(NodeUtils.PARAM_SSHPASSWORD, remotepassword, getSupplier(ssha, ssha::getPassword));
+            } else {
+                parameters.insert(NodeUtils.PARAM_SSHKEYFILE, sshkeyfile, getSupplier(ssha, ssha::getKeyfile));
+                parameters.insert(NodeUtils.PARAM_SSHKEYPASSPHRASE, //
```
[the remark this answers is not in the document]
Comment (dmatej) [reply nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeRemoteCommand.java]: Disagree. This prevents random reformatting to less readable code and it is still less annoying than using `// @formatter:off` (supported by Eclipse and Idea, but ugly)

**Comment by dmatej** [comment]
> Admin GUI seems to work nicely now.
> 
> It might be worth hiding the new `--sshauthtype` parameter for the asadmin commands, since it's not really usable from the command line - you aren't allowed to specify a password or alias from the command line so essentially all the parameter does is wipe your config joy

I experimented with it, and I even found one use case where it was useful - again - to reconfigure authentication to use defaults keyfile. Otherwise you cannot send null via command line, it is an empty string and it will be saved to domain.xml (and the behavior is same then).
So it might be useful.

Same with password, you can set password, but you have to provide passwordfile. If not, it will use user's private key as a default.

Seems it would be possible to use even passphrase again, but that goes behind borders of this issue.

> 
> Speaking of which, I see that it clears the key and password config when you switch between them.
> Just double-checking - this isn't going to trip anyone up is it? To me it makes sense but I don't know if there's any sort of "fallback" behaviour typically expected from SSH.

By default those values are clean => it uses system defaults (or trilead defaults? I'm not sure, but it tried to use my private key and in domain.xml was nothing).

Example (created node in description of this PR, I updated only the hostname):
```
    <node name="cmd-ssh001" install-dir="${com.sun.aas.productRoot}" type="SSH" node-host="localhost">
      <ssh-connector>
        <ssh-auth></ssh-auth>
      </ssh-connector>
    </node>
```

And "ping" result:
`Failed to validate SSH connection to node cmd-ssh001 (localhost) Could not connect to host localhost using SSH. Could not authenticate. SSH authentication with key file /home/dmatej/.ssh/id_rsa failed: PEM problem: it is of unknown type`

- the reason is that as expected, my private key has passphrase, so Trilead cannot detect even type of the key.

File: nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeRemoteCommand.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sgflt) [reply nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeRemoteCommand.java]: Better to properly configure formatter than this kind of hacky comments. We have used this too, but it is just a pain in the... if you have to refactor this "prettyfied" code.

[the remark this answers is not in the document]
Comment (dmatej) [reply nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeRemoteCommand.java]: Some formatting cannot be done automatically. I don't see any possible issue with refactoring.
I assume you only configured your editor to NOT formatting. Or tell me more, better on chat :)

[the remark this answers is not in the document]
Comment (dmatej) [reply nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/UpdateNodeRemoteCommand.java]: Compromise: static imports -> no need for formatting.

**Comment by dmatej** [comment]
Jenkins test please

**Comment by Pandrex247** [comment]
`So it might be useful.`
Ok, but we could still probably do with some improvements here - it provides no command-line help on what the acceptable values are and they're case-sensitive.

[Param.acceptableValues](https://github.com/payara/Payara/blob/master/nucleus/common/glassfish-api/src/main/java/org/glassfish/api/Param.java#L82) should help with the former if you didn't know about it.

**Comment by dmatej** [comment]
Jenkins test please

**Comment by dmatej** [comment]
Jenkins test please

**Pull request merged**